x-pack/plugins/stack_alerts/server/rule_types/es_query/executor.ts

```diff
@@ -114,13 +127,28 @@ export async function executor(core: CoreSetup, options: ExecutorOptions<EsQuery
       params,
       ...(isGroupAgg ? { group: alertId } : {}),
     });
-    const alert = alertFactory.create(
-      alertId === UngroupedGroupId && !isGroupAgg ? ConditionMetAlertInstanceId : alertId
-    );
-    alert
-      // store the params we would need to recreate the query that led to this alert instance
-      .replaceState({ latestTimestamp, dateStart, dateEnd })
-      .scheduleActions(ActionGroupId, actionContext);
+
+    const id = alertId === UngroupedGroupId && !isGroupAgg ? ConditionMetAlertInstanceId : alertId;
+
+    alertsClient!.report({
+      id,
+      actionGroup: ActionGroupId,
+      state: { latestTimestamp, dateStart, dateEnd },
+      context: actionContext,
+      payload: {
+        [ALERT_RULE_NAME]: name,
+        [ALERT_URL]: actionContext.link,
+        [ALERT_HITS_COUNT]: actionContext.hits.length,
+        [ALERT_HITS_HITS]: actionContext.hits,
+        [ALERT_MESSAGE]: actionContext.message,
+        [ALERT_TITLE]: actionContext.title,
+        [ALERT_CONDITIONS]: actionContext.conditions,
+        [ALERT_CONDITIONS_MET_VALUE]: actionContext.value,
+        [ALERT_STATE_LAST_TIMESTAMP]: latestTimestamp,
+        [ALERT_STATE_DATE_START]: dateStart,
+        [ALERT_STATE_DATE_END]: dateEnd,
+      },
+    });
```

Comment:
ES Query Rule used to pass context and state params to the legacy alertsClient. Now we pass them in the payload to save in alert-as-data index.

Fields in the state object: `dateEnd`, `dateStart`, and `latestTimestamp`. 
They are grouped under state namespace, like: `kibana.alert.state.latestTimestamp` 

Fields in the context object: `hits`, `link`, `conditions`,  `value`, `date`,  `title`, `message`. 
I didn't use any namespace for the context fields as they already make sense under the alert namespace.

Two field names has been created for the `hits` param.
`kibana.alert.hits.count` to pass the total number of the hits (in case of we want to cap the number of the hits)
`kibana.alert.hits.hits` to pass the hits (capped or all)

`ALERT_URL` (kibana.alert.url) is used from the default alert-as-data for the `link` param.

`ALERT_CONDITIONS_MET_VALUE` (kibana.alert.conditions_met_value) has been created for the `value` param in the context obj.

Do these field names and namespaces look OK to you?  Should we group context params as well?
Is creating a separate hits.count field name for the `hits` OK?

cc: @kobelb @shanisagiv1 @mikecote @ymao1

Reply:
I would put `actionContext.message` into `kibana.alert.reason` and look into the `ALERT_EVALUATION_THRESHOLD` and `ALERT_EVALUATION_VALUES` fields that are currently used by observability alerts to see if we could use them to store `actionContext.conditions` and `actionContext.value`. If they are suitable, we could promote those to framework level fields and encourage rule types to use them to store their evaluation thresholds.

For this first iteration, I would leave out `actionContext.hits` until we settle on how this data should be properly stored. I believe @kobelb had the idea of maybe copying out all ECS source fields

The other thing to think about is whether we want to use the `kibana.alerts` namespace directly or namespace these fields even more to be for stack alerts, so `kibana.alerts.stack_context.latest_timestamp` for example

Reply:
+1 to what @ymao1 mentioned. I've also been thinking the following:

### ALERT_HITS_COUNT
There may be more hits that provided within `actionContext.hits.length`. Should we instead adapt `fetchEsQuery` and `fetchSearchSourceQuery` to return `numMatches` and use that as the value? For example, setting size: 0 in the query will cause no hits, but ALERT_HITS_COUNT could contain `result.hits.total` from the query?

### ALERT_STATE_LAST_TIMESTAMP, ALERT_STATE_DATE_START, ALERT_STATE_DATE_END
I wonder if we leave these out for now? They are used for internal query building (however available in state.* mustache variables), I wonder if they provide value in the alert document given it would only represent information about the last rule run the alert was detected in (since we'd be constantly updating these fields each rule run but we have alert.start and alert.end timestamps already stored..)

Reply:
Currant fields are: 
```
[ALERT_RULE_NAME]: name,
[ALERT_URL]: actionContext.link,
[ALERT_HITS_COUNT]: actionContext.hits.length,
[ALERT_HITS_HITS]: actionContext.hits,
[ALERT_MESSAGE]: actionContext.message,
[ALERT_TITLE]: actionContext.title,
[ALERT_CONDITIONS]: actionContext.conditions,
[ALERT_CONDITIONS_MET_VALUE]: actionContext.value,
[ALERT_STATE_LAST_TIMESTAMP]: latestTimestamp,
[ALERT_STATE_DATE_START]: dateStart,
[ALERT_STATE_DATE_END]: dateEnd,
```        

I realised that `ALERT_RULE_NAME` is not in the current context, so i'll remove it.

I'll remove `ALERT_HITS_HITS` for now, as Ying suggested.

Will replace `ALERT_MESSAGE` with `ALERT_REASON`

`ALERT_EVALUATION_THRESHOLD` is used as a number, but `actionContext.conditions` in Es Query rule is the whole condition string. I'd rather adding `ALERT_EVALUATION_CONDITIONS` so it wold be compatible with the other `ALERT_EVALUATION_*`

I'll use `ALERT_EVALUATION_VALUES ` instead of `ALERT_CONDITIONS_MET_VALUE` (CONDITIONS and VALUES will be under the same namespace)

I also realised that `ALERT_HITS_COUNT` is the same thing as `ALERT_EVALUATION_VALUES`, so i'll remove it too.

And will remove all the state fields as Mike suggests.

So the new payload will be:  
```
[ALERT_URL]: actionContext.link,
[ALERT_REASON]: actionContext.message,
[ALERT_TITLE]: actionContext.title,
[ALERT_EVALUATION_CONDITIONS]: actionContext.conditions,
[ALERT_EVALUATION_VALUE]: actionContext.value,
```
And `ALERT_TITLE` will be the only ES Query specific field. Maybe we can move it to framework level as well as it sounds like a generic field name.

Reply:
Can you elaborate on what the conditions field will store? The type just has it being a string, is it a human readable string?

Otherwise, the revised fields look good to me.

Reply:
Gotcha, thanks Ersin. These look good to me.